### PR TITLE
feat: workspace context menu and picker menu

### PR DESCRIPTION
**Pull request by yuye-aws**
## Screenshot

image

image

image

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/public/chrome/constants.ts [hunk 1]
```diff
@@ -31,3 +31,7 @@
 export const OPENSEARCH_DASHBOARDS_ASK_OPENSEARCH_LINK = 'https://forum.opensearch.org/';
 export const GITHUB_CREATE_ISSUE_LINK =
   'https://github.com/opensearch-project/OpenSearch-Dashboards/issues/new/choose';
+
+export const WORKSPACE_CREATE_APP_ID = 'workspace_create';
+export const WORKSPACE_LIST_APP_ID = 'workspace_list';
+export const WORKSPACE_OVERVIEW_APP_ID = 'workspace_overview';
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/constants.ts:37]: There are duplicate constants in Workspace service, we should move these constants to a common directory.

File: src/core/public/chrome/ui/header/collapsible_nav.tsx [hunk 2]
```diff
@@ -215,6 +221,12 @@ export function CollapsibleNav({
       outsideClickCloses={false}
     >
       <EuiFlexItem className="eui-yScroll">
+        <CollapsibleNavHeader
+          getUrlForApp={getUrlForApp}
+          workspaces={workspaces}
+          basePath={basePath}
+        />
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav.tsx:228]: I suppose we can use `useOpensearchDashboards` to get all the props that is needed in CollapsivleNavHeader. In that way CollapsibleNav do not have to depend on these props.

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+export function CollapsibleNavHeader({ workspaces, getUrlForApp, basePath }: Props) {
+  const workspaceEnabled = useObservable(workspaces.workspaceEnabled$, false);
+  const workspaceList = useObservable(workspaces.workspaceList$, []);
+  const currentWorkspace = useObservable(workspaces.currentWorkspace$, null);
+  const filteredWorkspaceList = getFilteredWorkspaceList(
+    workspaces,
+    workspaceList,
+    currentWorkspace
+  );
+  const defaultHeaderName = 'OpenSearch Analytics';
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav_header.tsx:67]: We'd better wrap with i18n

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  };
+
+  const getWorkspaceListItems = () => {
+    const workspaceListItems = filteredWorkspaceList.map((workspace, index) =>
+      workspaceToItem(workspace, index)
+    );
+    const length = workspaceListItems.length;
+    workspaceListItems.push({
+      icon: <EuiIcon type="plus" />,
+      name: 'Create workspaces',
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav_header.tsx:122]: wrap with i18n

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        getUrlForApp(WORKSPACE_CREATE_APP_ID, {
+          absolute: false,
+        }),
+        currentWorkspace?.id ?? '',
+        basePath
+      ),
+    });
+    workspaceListItems.push({
+      icon: <EuiIcon type="folderClosed" />,
+      name: 'All workspaces',
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav_header.tsx:134]: i18n

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    </EuiFlexGroup>
+  );
+
+  const panels = [
+    {
+      id: 0,
+      title: currentWorkspaceTitle,
+      items: [
+        {
+          name: <EuiText style={{ fontWeight: 'bold' }}>{'Workspaces'}</EuiText>,
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav_header.tsx:183]: i18n

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+          panel: 1,
+        },
+        {
+          name: 'Management',
+          icon: 'managementApp',
+          href: formatUrlWithWorkspaceId(
+            getUrlForApp(WORKSPACE_OVERVIEW_APP_ID, {
+              absolute: false,
+            }),
+            'management',
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav_header.tsx:194]: There is a Management workspace id constant, please use that.

File: src/core/public/chrome/ui/header/nav_link.tsx [hunk 8]
```diff
@@ -73,7 +73,6 @@ export function createEuiListItem({
       }
 
       if (
-        !link.externalLink && // ignore external links
```
Comment (wanglam) [inline src/core/public/chrome/ui/header/nav_link.tsx:76]: Is this change by intention? Seems the external link will pass here.

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    // current workspace located at the top of workspace list
+    filteredWorkspaceList = filteredWorkspaceList.filter(
+      (workspace) => workspace.id !== currentWorkspace.id
+    );
+    filteredWorkspaceList.unshift(currentWorkspace);
+    if (filteredWorkspaceList.length > 5) {
+      filteredWorkspaceList.pop();
+    }
+  }
+  return filteredWorkspaceList;
```
Comment (wanglam) [inline src/core/public/chrome/ui/header/collapsible_nav_header.tsx:55]: Prefer refactor code like
```tsx
return [
    ...(currentWorkspace ? [currentWorkspace]:[]),
   ...workspaceList
    .filter((workspace) => workspace.id !== MANAGEMENT_WORKSPACE && workspace.id !== currentWorkspace?.id)
].slice(0, 5);
```

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+export function CollapsibleNavHeader({ workspaces, getUrlForApp, basePath }: Props) {
```
Comment (wanglam) [inline src/core/public/chrome/ui/header/collapsible_nav_header.tsx:58]: @SuZhou-Joe @ruanyl Do we need to move these workspace related code to workspace plugin?

File: src/core/public/chrome/constants.ts (hunk 1 above)
Comment (ruanyl) [reply src/core/public/chrome/constants.ts:37]: @yuye-aws and I discussed this and agreed on this implementation, but after thinking it more throughly, I think this is not ideal. Core app should not aware what's workspace create/list/overview page. 
@SuZhou-Joe Do you think if we can leave it like this for now, and I will make another PR to refactor it.

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx (hunk 10 above)
Comment (ruanyl) [reply src/core/public/chrome/ui/header/collapsible_nav_header.tsx:58]: @wanglam Any suggestion on how to render it in left menu if moved to plugin folder?

Comment (ruanyl) [reply src/core/public/chrome/ui/header/collapsible_nav_header.tsx:58]: Maybe something like this:

```typescript
// src/core/public/workspace/workspaces_service.ts
class WorkspaceService {
  private _renderSelector;

  renderSelector() {
    if (this._renderSelector) {
      this._renderSelector()
    }
  }

  onRenderSelector(render) {
    this._renderSelector = render
  }
}

// src/plugins/workspace/public/plugin.ts
class WorkspacePlugin {
  setup(core) {
    core.workspaces.onRenderSelector(mountSelector)
  }
}

function mountSelector() {
  return <CollapsibleNavHeader />
}
```
Does it look good to you?

File: src/core/public/chrome/ui/header/nav_link.tsx (hunk 8 above)
Comment (yuye-aws) [reply src/core/public/chrome/ui/header/nav_link.tsx:76]: Yes. `link.externalLink` logic was previously added to enable workspace switch in left menu, which is no longer needed.

File: src/core/public/chrome/constants.ts (hunk 1 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/constants.ts:37]: Sure, see the proposal that you want to move the render logic into plugin/workspace.

File: src/core/public/chrome/ui/header/collapsible_nav_header.tsx (hunk 10 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/ui/header/collapsible_nav_header.tsx:51]: Looks good to me.

Comment (yuye-aws) [reply src/core/public/chrome/ui/header/collapsible_nav_header.tsx:51]: Should this change be included in this PR?

**Review by ruanyl — APPROVED** [review]

**Review by wanglam — APPROVED** [review]

**Pull request merged**